File: forge-core/internal/kerneldecisioncontract/graph.go
package kerneldecisioncontract

import (
	"encoding/json"
	"fmt"
	"math"
	"reflect"

	op "forgeos/forge-core/internal/kerneloperationalcontract"
)

func atomIndex(values []CognitiveAtom) map[string]*CognitiveAtom {
	result := make(map[string]*CognitiveAtom, len(values))
	for index := range values {
		result[values[index].AtomID] = &values[index]
	}
	return result
}

func transactionRefs(value *DecisionTransaction) []AtomRef {
	result := []AtomRef{value.GoalAtomRef}
	result = append(result, value.TriggerAtomRefs...)
	return append(result, value.GuardAtomRefs...)
}

func validateRoleClosure(atoms []CognitiveAtom, transaction *DecisionTransaction) error {
	index := atomIndex(atoms)
	referenced := make(map[string]bool)
	for roleIndex, reference := range transactionRefs(transaction) {
		atom := index[reference.AtomID]
		if atom == nil || atom.AtomSHA256 != reference.AtomSHA256 {
			return fmt.Errorf("transaction atom reference does not resolve exact CognitiveAtom")
		}
		if !oneOf(atom.Source.SourceKind, "artifact", "cognitive_atom_v1", "evidence_record", "work_intent") {
			return fmt.Errorf("DecisionTransaction may reference only predecision atoms")
		}
		if (roleIndex == 0) != (atom.AtomType == "goal") {
			return fmt.Errorf("goal_atom_ref must resolve the only predecision goal CognitiveAtom")
		}
		referenced[reference.AtomID] = true
	}
	for index := range atoms {
		pre := atoms[index].Source.SourcePhase == "predecision"
		if pre != referenced[atoms[index].AtomID] {
			return fmt.Errorf("predecision atoms must equal exact transaction role union")
		}
	}
	return nil
}

func rawReference(raw json.RawMessage) (string, string, error) {
	var value map[string]any
	if err := json.Unmarshal(raw, &value); err != nil {
		return "", "", err
	}
	var id, digest string
	for field, member := range value {
		textValue, ok := member.(string)
		if !ok {
			return "", "", fmt.Errorf("operational source reference is not textual")
		}
		if len(field) >= 3 && field[len(field)-3:] == "_id" {
			id = textValue
		}
		if len(field) >= 7 && field[len(field)-7:] == "_sha256" {
			digest = textValue
		}
	}
	if id == "" || digest == "" {
		return "", "", fmt.Errorf("operational source reference fields are missing")
	}
	return id, digest, nil
}

func validatePostSources(atoms []CognitiveAtom, transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	for index := range atoms {
		atom := &atoms[index]
		if atom.Source.SourcePhase != "postdecision" {
			continue
		}
		id, digest, err := rawReference(atom.Source.SourceRef)
		if err != nil {
			return err
		}
		time, output, found := operationalSource(operational, atom.Source.SourceKind, id, digest)
		if !found {
			return fmt.Errorf("postdecision source_ref must resolve exact operational record")
		}
		if atom.Source.SourceKind == "artifact_receipt" && !output {
			return fmt.Errorf("postdecision ArtifactReceipt source must be declared_output")
		}
		if atom.Validity.ValidFromUnixMS < transaction.CreatedAtUnixMS ||
			atom.Validity.ValidFromUnixMS < time {
			return fmt.Errorf("postdecision Atom validity predates transaction or source")
		}
	}
	return nil
}

func operationalSource(value *op.KernelOperationalReferenceClosure, kind, id, digest string) (int64, bool, bool) {
	switch kind {
	case "artifact_receipt":
		for _, record := range value.ArtifactReceipts {
			if record.ArtifactReceiptID == id && record.ArtifactReceiptSHA256 == digest {
				return record.CreatedAtUnixMS, record.ReceiptRole == "declared_output", true
			}
		}
	case "capability_invocation":
		for _, record := range value.CapabilityInvocations {
			if record.InvocationID == id && record.InvocationSHA256 == digest {
				return record.RequestedAtUnixMS, false, true
			}
		}
	case "interaction_event":
		for _, record := range value.InteractionEvents {
			if record.EventID == id && record.EventSHA256 == digest {
				return record.OccurredAtUnixMS, false, true
			}
		}
	case "execution_receipt":
		for _, record := range value.ExecutionReceipts {
			if record.ExecutionReceiptID == id && record.ExecutionReceiptSHA256 == digest {
				return record.EndedAtUnixMS, false, true
			}
		}
	}
	return 0, false, false
}

func validateContext(atoms []CognitiveAtom, transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	for index := range atoms {
		if !reflect.DeepEqual(atoms[index].TaskBinding, transaction.TaskBinding) ||
			!reflect.DeepEqual(atoms[index].Bindings, transaction.Bindings) {
			return fmt.Errorf("every CognitiveAtom must share transaction task and bindings")
		}
	}
	for _, record := range operational.ArtifactReceipts {
		if !reflect.DeepEqual(record.TaskBinding, transaction.TaskBinding) || record.Bindings != transaction.Bindings {
			return fmt.Errorf("ArtifactReceipt task or bindings drift")
		}
	}
	for _, record := range operational.CapabilityInvocations {
		if !reflect.DeepEqual(record.TaskBinding, transaction.TaskBinding) || record.Bindings != transaction.Bindings {
			return fmt.Errorf("CapabilityInvocation task or bindings drift")
		}
	}
	return validateRemainingContext(transaction, operational)
}

func validateRemainingContext(transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	for _, record := range operational.InteractionEvents {
		if !reflect.DeepEqual(record.TaskBinding, transaction.TaskBinding) || record.Bindings != transaction.Bindings {
			return fmt.Errorf("InteractionEvent task or bindings drift")
		}
	}
	for _, record := range operational.ExecutionReceipts {
		if !reflect.DeepEqual(record.TaskBinding, transaction.TaskBinding) || record.Bindings != transaction.Bindings {
			return fmt.Errorf("ExecutionReceipt task or bindings drift")
		}
	}
	return nil
}

func selectedOption(value *DecisionTransaction) *DecisionOption {
	for index := range value.Options {
		if value.Options[index].OptionID == value.SelectedOptionID {
			return &value.Options[index]
		}
	}
	return nil
}

func validateSelectedOperation(transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	selected := selectedOption(transaction)
	for _, invocation := range operational.CapabilityInvocations {
		if invocation.CorrelationID != transaction.DecisionTransactionID ||
			invocation.Subject != transaction.Actor || invocation.Capability != selected.Capability ||
			invocation.RequestedActionSHA256 != selected.RequestedActionSHA256 ||
			invocation.IdempotencyKey != transaction.IdempotencyKey ||
			!reflect.DeepEqual(invocation.InputArtifactReceiptRefs, transaction.ReadArtifactReceiptRefs) ||
			!reflect.DeepEqual(invocation.DeclaredOutputSlots, transaction.WriteSlots) {
			return fmt.Errorf("Invocation differs from selected transaction declarations")
		}
	}
	for _, event := range operational.InteractionEvents {
		if event.CorrelationID != transaction.DecisionTransactionID {
			return fmt.Errorf("InteractionEvent correlation differs from transaction")
		}
	}
	for _, receipt := range operational.ExecutionReceipts {
		if receipt.CorrelationID != transaction.DecisionTransactionID {
			return fmt.Errorf("ExecutionReceipt correlation differs from transaction")
		}
	}
	return nil
}

func validateTimes(atoms []CognitiveAtom, transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	first := operational.CapabilityInvocations[0].RequestedAtUnixMS
	for _, invocation := range operational.CapabilityInvocations[1:] {
		if invocation.RequestedAtUnixMS < first {
			first = invocation.RequestedAtUnixMS
		}
	}
	if transaction.CreatedAtUnixMS > first {
		return fmt.Errorf("transaction creation follows first request")
	}
	if err := validateReadTimes(transaction, operational); err != nil {
		return err
	}
	for index := range atoms {
		if atoms[index].Source.SourcePhase != "predecision" {
			continue
		}
		validity := atoms[index].Validity
		if validity.ValidFromUnixMS > transaction.CreatedAtUnixMS ||
			validity.ValidUntilUnixMS != nil && transaction.CreatedAtUnixMS >= *validity.ValidUntilUnixMS {
			return fmt.Errorf("predecision Atom is future or expired at transaction creation")
		}
	}
	return nil
}

func validateReadTimes(transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	for _, reference := range transaction.ReadArtifactReceiptRefs {
		found := false
		for _, receipt := range operational.ArtifactReceipts {
			if receipt.ArtifactReceiptID == reference.ArtifactReceiptID &&
				receipt.ArtifactReceiptSHA256 == reference.ArtifactReceiptSHA256 &&
				receipt.ReceiptRole == "declared_input" &&
				receipt.CreatedAtUnixMS <= transaction.CreatedAtUnixMS {
				found = true
			}
		}
		if !found {
			return fmt.Errorf("transaction read must resolve nonfuture declared-input receipt")
		}
	}
	return nil
}

func checkedAdd(total, increment int64) (int64, error) {
	if increment < 0 || total > math.MaxInt64-increment {
		return 0, fmt.Errorf("caller-declared aggregate usage overflows signed int64")
	}
	return total + increment, nil
}

func validateBudgetClosure(transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	if int64(len(operational.CapabilityInvocations)) > transaction.Budget.MaxCalls {
		return fmt.Errorf("Invocation count exceeds transaction budget")
	}
	totals := [7]int64{}
	for _, receipt := range operational.ExecutionReceipts {
		usage := receipt.ObservedUsage
		values := [7]int64{usage.CallCount, usage.CostUSDMicros, usage.ElapsedMS,
			usage.InputTokens, usage.NetworkBytes, usage.OutputBytes, usage.OutputTokens}
		for index, increment := range values {
			var err error
			totals[index], err = checkedAdd(totals[index], increment)
			if err != nil {
				return err
			}
		}
	}
	limits := [7]int64{transaction.Budget.MaxCalls, transaction.Budget.MaxCostUSDMicros,
		transaction.Budget.TimeoutMS, transaction.Budget.MaxInputTokens,
		transaction.Budget.MaxNetworkBytes, transaction.Budget.MaxOutputBytes,
		transaction.Budget.MaxOutputTokens}
	for index := range totals {
		if totals[index] > limits[index] {
			return fmt.Errorf("caller-declared aggregate usage exceeds transaction budget")
		}
	}
	return nil
}

func validateReferenceGraph(atoms []CognitiveAtom, transaction *DecisionTransaction,
	operational *op.KernelOperationalReferenceClosure) error {
	if err := validateRoleClosure(atoms, transaction); err != nil {
		return err
	}
	if err := validatePostSources(atoms, transaction, operational); err != nil {
		return err
	}
	if err := validateContext(atoms, transaction, operational); err != nil {
		return err
	}
	if err := validateSelectedOperation(transaction, operational); err != nil {
		return err
	}
	if err := validateTimes(atoms, transaction, operational); err != nil {
		return err
	}
	return validateBudgetClosure(transaction, operational)
}

File: forge-runtime/crates/domain/src/kernel_decision_contract/graph.rs
use std::collections::{BTreeMap, BTreeSet};

use crate::kernel_operational_contract::{KernelOperationalReferenceClosure, ObservedUsage};

use super::{
    CognitiveAtom, DecisionOption, DecisionTransaction, KernelDecisionContractError, invalid,
    source::raw_reference,
};

pub(super) fn role_closure(
    atoms: &[CognitiveAtom],
    transaction: &DecisionTransaction,
) -> Result<(), KernelDecisionContractError> {
    let index = atoms
        .iter()
        .map(|atom| (atom.atom_id.as_str(), atom))
        .collect::<BTreeMap<_, _>>();
    let role_refs = std::iter::once(&transaction.goal_atom_ref)
        .chain(&transaction.trigger_atom_refs)
        .chain(&transaction.guard_atom_refs)
        .collect::<Vec<_>>();
    let mut role_ids = BTreeSet::new();
    for (role_index, reference) in role_refs.into_iter().enumerate() {
        let atom = index
            .get(reference.atom_id.as_str())
            .ok_or_else(|| invalid("transaction references a missing CognitiveAtom"))?;
        if atom.atom_sha256 != reference.atom_sha256 {
            return Err(invalid("transaction CognitiveAtom digest drift"));
        }
        if !matches!(
            atom.source.source_kind.as_str(),
            "artifact" | "cognitive_atom_v1" | "evidence_record" | "work_intent"
        ) {
            return Err(invalid(
                "DecisionTransaction may reference only predecision atoms",
            ));
        }
        if (role_index == 0) != (atom.atom_type == "goal") {
            return Err(invalid(
                "goal_atom_ref must resolve the only predecision goal CognitiveAtom",
            ));
        }
        role_ids.insert(reference.atom_id.as_str());
    }
    let predecision = atoms
        .iter()
        .filter(|atom| atom.source.source_phase == "predecision")
        .map(|atom| atom.atom_id.as_str())
        .collect::<BTreeSet<_>>();
    if predecision != role_ids {
        return Err(invalid(
            "predecision atoms must equal exact transaction role union",
        ));
    }
    Ok(())
}

fn operational_source(
    value: &KernelOperationalReferenceClosure,
    kind: &str,
    id: &str,
    digest: &str,
) -> Option<(i64, bool)> {
    match kind {
        "artifact_receipt" => value
            .artifact_receipts
            .iter()
            .find(|record| {
                record.artifact_receipt_id == id && record.artifact_receipt_sha256 == digest
            })
            .map(|record| {
                (
                    record.created_at_unix_ms,
                    record.receipt_role == "declared_output",
                )
            }),
        "capability_invocation" => value
            .capability_invocations
            .iter()
            .find(|record| record.invocation_id == id && record.invocation_sha256 == digest)
            .map(|record| (record.requested_at_unix_ms, false)),
        "interaction_event" => value
            .interaction_events
            .iter()
            .find(|record| record.event_id == id && record.event_sha256 == digest)
            .map(|record| (record.occurred_at_unix_ms, false)),
        "execution_receipt" => value
            .execution_receipts
            .iter()
            .find(|record| {
                record.execution_receipt_id == id && record.execution_receipt_sha256 == digest
            })
            .map(|record| (record.ended_at_unix_ms, false)),
        _ => None,
    }
}

fn post_sources(
    atoms: &[CognitiveAtom],
    transaction: &DecisionTransaction,
    operational: &KernelOperationalReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    for atom in atoms {
        if atom.source.source_phase != "postdecision" {
            continue;
        }
        let (id, digest) = raw_reference(&atom.source.source_ref)?;
        let (source_time, declared_output) =
            operational_source(operational, &atom.source.source_kind, id, digest)
                .ok_or_else(|| invalid("postdecision source_ref does not resolve exact record"))?;
        if atom.source.source_kind == "artifact_receipt" && !declared_output {
            return Err(invalid(
                "postdecision ArtifactReceipt source must be declared_output",
            ));
        }
        if atom.validity.valid_from_unix_ms < transaction.created_at_unix_ms
            || atom.validity.valid_from_unix_ms < source_time
        {
            return Err(invalid(
                "postdecision Atom validity predates transaction or source",
            ));
        }
    }
    Ok(())
}

fn context(
    atoms: &[CognitiveAtom],
    transaction: &DecisionTransaction,
    operational: &KernelOperationalReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    if atoms.iter().any(|atom| {
        atom.task_binding != transaction.task_binding || atom.bindings != transaction.bindings
    }) {
        return Err(invalid(
            "every CognitiveAtom must share transaction task and bindings",
        ));
    }
    let receipt_drift = operational.artifact_receipts.iter().any(|record| {
        record.task_binding != transaction.task_binding || record.bindings != transaction.bindings
    });
    let invocation_drift = operational.capability_invocations.iter().any(|record| {
        record.task_binding != transaction.task_binding || record.bindings != transaction.bindings
    });
    let event_drift = operational.interaction_events.iter().any(|record| {
        record.task_binding != transaction.task_binding || record.bindings != transaction.bindings
    });
    let execution_drift = operational.execution_receipts.iter().any(|record| {
        record.task_binding != transaction.task_binding || record.bindings != transaction.bindings
    });
    if receipt_drift || invocation_drift || event_drift || execution_drift {
        return Err(invalid("operational record task or bindings drift"));
    }
    Ok(())
}

fn selected(value: &DecisionTransaction) -> &DecisionOption {
    value
        .options
        .iter()
        .find(|option| option.option_id == value.selected_option_id)
        .expect("validated selected option")
}

fn selected_operation(
    transaction: &DecisionTransaction,
    operational: &KernelOperationalReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    let option = selected(transaction);
    for invocation in &operational.capability_invocations {
        if invocation.correlation_id != transaction.decision_transaction_id
            || invocation.subject != transaction.actor
            || invocation.capability != option.capability
            || invocation.requested_action_sha256 != option.requested_action_sha256
            || invocation.idempotency_key != transaction.idempotency_key
            || invocation.input_artifact_receipt_refs != transaction.read_artifact_receipt_refs
            || invocation.declared_output_slots != transaction.write_slots
        {
            return Err(invalid(
                "Invocation differs from selected transaction declarations",
            ));
        }
    }
    if operational
        .interaction_events
        .iter()
        .any(|event| event.correlation_id != transaction.decision_transaction_id)
        || operational
            .execution_receipts
            .iter()
            .any(|receipt| receipt.correlation_id != transaction.decision_transaction_id)
    {
        return Err(invalid("operational correlation differs from transaction"));
    }
    Ok(())
}

pub(super) fn times(
    atoms: &[CognitiveAtom],
    transaction: &DecisionTransaction,
    operational: &KernelOperationalReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    let first = operational
        .capability_invocations
        .iter()
        .map(|value| value.requested_at_unix_ms)
        .min()
        .expect("operational closure has invocation");
    if transaction.created_at_unix_ms > first {
        return Err(invalid("transaction creation follows first request"));
    }
    read_times(transaction, operational)?;
    if atoms.iter().any(|atom| {
        atom.source.source_phase == "predecision"
            && (atom.validity.valid_from_unix_ms > transaction.created_at_unix_ms
                || atom
                    .validity
                    .valid_until_unix_ms
                    .is_some_and(|end| transaction.created_at_unix_ms >= end))
    }) {
        return Err(invalid(
            "predecision Atom is future or expired at transaction creation",
        ));
    }
    Ok(())
}

fn read_times(
    transaction: &DecisionTransaction,
    operational: &KernelOperationalReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    for reference in &transaction.read_artifact_receipt_refs {
        let found = operational.artifact_receipts.iter().any(|receipt| {
            receipt.artifact_receipt_id == reference.artifact_receipt_id
                && receipt.artifact_receipt_sha256 == reference.artifact_receipt_sha256
                && receipt.receipt_role == "declared_input"
                && receipt.created_at_unix_ms <= transaction.created_at_unix_ms
        });
        if !found {
            return Err(invalid(
                "transaction read must resolve nonfuture declared-input receipt",
            ));
        }
    }
    Ok(())
}

fn usage_values(value: &ObservedUsage) -> [i64; 7] {
    [
        value.call_count,
        value.cost_usd_micros,
        value.elapsed_ms,
        value.input_tokens,
        value.network_bytes,
        value.output_bytes,
        value.output_tokens,
    ]
}

pub(super) fn checked_add_usage(
    total: i64,
    increment: i64,
) -> Result<i64, KernelDecisionContractError> {
    total
        .checked_add(increment)
        .ok_or_else(|| invalid("caller-declared aggregate usage overflows signed int64"))
}

pub(super) fn budget_closure(
    transaction: &DecisionTransaction,
    operational: &KernelOperationalReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    if i64::try_from(operational.capability_invocations.len()).unwrap_or(i64::MAX)
        > transaction.budget.max_calls
    {
        return Err(invalid("Invocation count exceeds transaction budget"));
    }
    let mut totals = [0_i64; 7];
    for receipt in &operational.execution_receipts {
        for (index, increment) in usage_values(&receipt.observed_usage)
            .into_iter()
            .enumerate()
        {
            totals[index] = checked_add_usage(totals[index], increment)?;
        }
    }
    let limits = [
        transaction.budget.max_calls,
        transaction.budget.max_cost_usd_micros,
        transaction.budget.timeout_ms,
        transaction.budget.max_input_tokens,
        transaction.budget.max_network_bytes,
        transaction.budget.max_output_bytes,
        transaction.budget.max_output_tokens,
    ];
    if totals
        .iter()
        .zip(limits)
        .any(|(total, limit)| *total > limit)
    {
        return Err(invalid(
            "caller-declared aggregate usage exceeds transaction budget",
        ));
    }
    Ok(())
}

pub(super) fn validate_reference_graph(
    atoms: &[CognitiveAtom],
    transaction: &DecisionTransaction,
    operational: &KernelOperationalReferenceClosure,
) -> Result<(), KernelDecisionContractError> {
    role_closure(atoms, transaction)?;
    post_sources(atoms, transaction, operational)?;
    context(atoms, transaction, operational)?;
    selected_operation(transaction, operational)?;
    times(atoms, transaction, operational)?;
    budget_closure(transaction, operational)
}
